### ISSUE TITLE
 acme/autocert: no supported challenge type found

### ISSUE THREAD

**Issue by ramon-ga** [centrifugal/centrifugo#210]
Hi
I'm getting the following error for auto cert, after a fresh setup on a new server:
`acme/autocert: no supported challenge type found`

Settings are:
```
    "ssl_autocert": true,
    "ssl_autocert_host_whitelist": "host1,host2",
    "ssl_autocert_cache_dir": "/opt/certs/",
    "ssl_autocert_email": "email",
    "ssl_autocert_force_rsa": true,
    "ssl_autocert_server_name": "host1",
```

Maybe something is changed in the acme library?

**Comment by ramon-ga** [comment]
Hi @FZambia 
quite possible this is related to that issue.
I briefly read it, and i think they released an update for `crypto/acme`
https://github.com/golang/go/issues/21890#issuecomment-357986015

It's possible that you check if the dependency is up to date?
Else i'll find time next week to build it and test it...

The setup is as before on the old server, and there's no proxy in front.

**Comment by FZambia** [comment]
Yes, I'll update dependency. Looks like this also requires HTTP handler that works on port 80 to be registered in server and only solves a task of acme http_01 challenging.

**Comment by FZambia** [comment]
@ramon-ga added new boolean option `ssl_autocert_http` in https://github.com/centrifugal/centrifugo/commit/4920a8f6c7fc529fa1c2faa9896a05a3a22b1c13 - could you try Centrifugo from master branch on your host? You can try without enabling that option first - maybe updating acme library is enough. Otherwise that option will turn on handling http_01 challenges on port 80.

**Comment by ramon-ga** [comment]
@FZambia thanks for your effort on this! I tried it, see below.

With `ssl_autocert_http:false`:
```
[I]: 2018/04/09 07:21:01 Start serving raw websocket, SockJS, API, admin, web endpoints on :443
[I]: 2018/04/09 07:21:25 http: TLS handshake error from 212.51.141.47:53095: acme/autocert: unable to authorize "****.****.**"; tried ["tls-sni-02" "tls-sni-01"]
[I]: 2018/04/09 07:21:26 http: TLS handshake error from 212.51.141.47:53096: acme/autocert: missing certificate
```

With `ssl_autocert_http:true` it's working fine.

Points to consider:
- Should be written in log, that listen on port 80 too
- Maybe in that case, admin panel should no be accessible on port 80
- For me it's nice that on port 443 and 80 the service is available, but not for everyone?

**Comment by FZambia** [comment]
ok, fine 👍 

> Should be written in log, that listen on port 80 too

Agree, will add it.

> Maybe in that case, admin panel should no be accessible on port 80

Could you write more about admin panel? As far as I understand it must work on main port which is 443 in this case and not on port 80 - or am I wrong?

> For me it's nice that on port 443 and 80 the service is available, but not for everyone

This is turned on only if `ssl_autocert_http` turned on and only serves acme challenges on that port redirecting all other plain HTTP requests to HTTPS. What's you suggestion here?

**Comment by ramon-ga** [comment]
> As far as I understand it must work on main port which is 443 in this case and not on port 80
Correct

> This is turned on only if ssl_autocert_http turned on and only serves acme challenges on that port redirecting all other plain HTTP requests to HTTPS. What's you suggestion here?
Perfect, redirecting all other plain HTTP requests to HTTPS is a perfect solution

**Comment by FZambia** [comment]
> Correct

Sorry - so do you see any problem with admin interface at moment?

**Comment by ramon-ga** [comment]
No i don't see any problem for me this works fine.

I'm thinking just about that maybe someone don't like to see that the admin interface is accessible over unencrypted http. But for me it's fine.
Now if `ssl_autocert_http` is enabled, the admin interface is accessible through http too.

**Comment by FZambia** [comment]
Are you sure it was not a redirect to https admin page?

**Comment by FZambia** [comment]
@ramon-ga added logging in latest commit, so looks like this can be released?

**Comment by ramon-ga** [comment]
All this works like a charm, thank you for your effort!

Tasks done:
- [x] pull latest commit
- [x] build
- [x] delete existing certificates
- [x] run with new config flag `ssl_autocert_http: true`
- [x] open web with http for both domains
- [x] will redirect to https
- [x] certs `Let's Encrypt Authority X3` received
- [x] admin interface with wss connected

Log:
```
[I]: 2018/04/13 07:15:33 Redis: 127.0.0.1:6379/0, pool: 256, using password: no, API enabled: no
[I]: 2018/04/13 07:15:33 Config path: /etc/centrifugo/config.json
[I]: 2018/04/13 07:15:33 Version: 
[I]: 2018/04/13 07:15:33 PID: 20774
[I]: 2018/04/13 07:15:33 Engine: Redis
[I]: 2018/04/13 07:15:33 GOMAXPROCS: 8
[I]: 2018/04/13 07:15:33 SockJS url: //cdn.jsdelivr.net/sockjs/1.1/sockjs.min.js
[I]: 2018/04/13 07:15:33 Start serving raw websocket, SockJS, API, admin, web endpoints on :443
[I]: 2018/04/13 07:15:33 Serving ACME http_01 challenge on :80
```

**Comment by FZambia** [comment]
Just released this in [1.7.9](https://github.com/centrifugal/centrifugo/releases/tag/v1.7.9)
